### ISSUE TITLE
Strange behavior from Metadata.sign() and Metadata.to_file()

### ISSUE THREAD

**Issue by MVrachev** [theupdateframework/tuf#1812]
**Description of issue or feature request**:
I made some tests locally to answer the question: do different orders in the dictionaries in the metadata objects lead to a different signature?

I wrote this small script aiming to change the order of the `Targets.target` dictionary content as the insertion order is preserved.

```python
from tuf.api.metadata import Metadata, Targets
from securesystemslib.signer import SSlibSigner
from securesystemslib.interface import import_ed25519_privatekey_from_file
from tuf.api.serialization.json import JSONSerializer
import copy

serializer = JSONSerializer(compact=False)
old_target = Metadata.from_file("<LOCAL_PATH_TO_TUF>/tests/repository_data/repository/metadata/targets.json")

targets_key = import_ed25519_privatekey_from_file("<LOCAL_PATH_TO_TUF>/tests/repository_data/keystore/targets_key", password="password")
targets_signer = SSlibSigner(targets_key)

new_target = copy.deepcopy(old_target)
if isinstance(new_target.signed, Targets) and isinstance(old_target.signed, Targets):
    new_target.signed.targets = {}
    new_target.signed.targets["file2.txt"] = old_target.signed.targets["file2.txt"]
    new_target.signed.targets["file1.txt"] = old_target.signed.targets["file1.txt"]
    print(new_target == old_target)
    new_target.sign(targets_signer)
    new_target.to_file("targets_new_file.json", serializer)


for key, val in new_target.signed.targets.items():
    print(f"{key}: {val}")
```

and I expected that I will get one of the two scenarios:
1) the `targets_new_file.json` will have the same signature as the original `targets.json`, but a different order in the `singed.targets` dictionary
2) the `targets_new_file.json` will have a different signature compared to the original `targets.json` and different order in the `singed.targets` dictionary

Instead, I got a third scenario:
the `targets_new_file.json` has the same signature as the original `targets.json` and the **SAME** order in the `singed.targets` dictionary.
The terminal output looks like this: 
```
True
file2.txt: <tuf.api.metadata.TargetFile object at 0x7f588e1c0070>
file1.txt: <tuf.api.metadata.TargetFile object at 0x7f588e21b7c0>
```
meaning the insertion order is indeed preserved during iteration.

I debug it locally and I found the following:
1) We have the following calls:
`new_target.sign(targets_signer)` -> `CanonicalJSONSerializer.serialize()` -> `securesystemslib.formats.encode_canonical()`
I noticed a problem inside `securesystemslib.formats.encode_canonical()`. 
https://github.com/secure-systems-lab/securesystemslib/blob/602d3fcd59f726fc64e78a93430f427712d26ca4/securesystemslib/formats.py#L646.
The problem is that no matter the dict order at line 646 inside `securesystemslib.formats` we are sorting it.
In my case I always observed that `new_targets.signed_targets` containing `file2` and `file1` was traversed as `file1` and then `file2`.
Should we fix `securesystemslib.formats.encode_canonical()`? 
This means that signing two files with the same content, but a different order of delegations will receive the same signature. Is that okay?
2) Not sure how that happens, but I see the same problem as 1) when we call `Metadata.to_file`.
No matter the `new_targets.targets` order I always saw `file1` was first and `file2` was second. This time there is no securesystemslib call that made me suspicious, but instead, the problem should be somewhere inside `Metadata.to_bytes()`.
While debugging I found that it's not the problem when calling `md_dict = metadata_obj.to_dict()` inside `JSONSerializer.serialize()` as the `md_dict` has the same order of the targets as the `new_targets` object.
So, I am not sure where is the problem here...

@jku or @lukpueh does anyone from you have an opinion?
I am sure this is hard to follow, but I didn't find a better way to put it and I didn't want to forget my findings.

**Comment by jku** [comment]
Great testing, we should be doing more of this sort of verification of assumptions.

My first thought is that you may have mixed delegated roles and targets (a mistake I've made before) in that within Targets metadata, former is ordered but the latter is not: targets order does not matter but delegated roles does.

I'm not totally sure why the targets dict order you chose isn't reflected in the serialized format... but `encode_canonical()` seems to work as expected: dicts like "targets" are sorted (because their order should not affect canonicalization) but lists like "delegated roles" are not sorted because the order matters.

**Comment by jku** [comment]
> I'm not totally sure why the targets dict order you chose isn't reflected in the serialized format

It's because we ask json to sort keys:
```
            json_bytes = json.dumps(
                metadata_obj.to_dict(),
                indent=indent,
                separators=separators,
                sort_keys=True,
            )
```
hmm... I wonder if we really want this.

**Comment by MVrachev** [comment]
You are correct that `Targets.targets` order is not important as compared to `Targets.delegations.roles`.

I changed my script to this:
```python
from tuf.api.metadata import Metadata, Delegations, Targets
from securesystemslib.signer import SSlibSigner
from securesystemslib.interface import import_ed25519_privatekey_from_file
from tuf.api.serialization.json import JSONSerializer
import copy

serializer = JSONSerializer(compact=False)

old_target = Metadata.from_file("<LOCAL_PATH_TO_TUF>/tests/repository_data/fishy_rolenames/metadata/1.targets.json")

targets_key = import_ed25519_privatekey_from_file("<LOCAL_PATH_TO_TUF>/tests/repository_data/keystore/targets_key", password="password")
targets_signer = SSlibSigner(targets_key)

new_target = copy.deepcopy(old_target)
if isinstance(new_target.signed, Targets) and isinstance(old_target.signed, Targets):

    assert isinstance(old_target.signed.delegations, Delegations)
    assert isinstance(new_target.signed.delegations, Delegations)
    new_target.signed.delegations.roles = {}
    # Change the order of the roles compared to the old_target
    tmp_store = []
    for role_name, role_info in old_target.signed.delegations.roles.items():
        tmp_store.append((role_name, role_info))

    index = len(tmp_store) - 1
    while index >= 0:
        role_name, role_info = tmp_store[index]
        new_target.signed.delegations.roles[role_name] = role_info
        index -= 1

    print(new_target == old_target)
    new_target.sign(targets_signer)
    new_target.to_file("targets_new_file.json", serializer)


for key, val in new_target.signed.targets.items():
    print(f"{key}: {val}")
```

This time I changed the delegations roles order and indeed managed to verify that: 
**the order is important when calculating the file hashes**. 
The signatures keyid is the same (meaning I have used the same key), but I was able to produce a different signature.

I think this confirms that when I work towards a solution for #1788 I must make sure that when comparing dictionaries the order IS taken into account.
No matter if the specification cares about a particular dictionary order or not (as in the case of `Targets.targets`).


> > I'm not totally sure why the targets dict order you chose isn't reflected in the serialized format
> 
> It's because we ask json to sort keys:
> 
> ```
>             json_bytes = json.dumps(
>                 metadata_obj.to_dict(),
>                 indent=indent,
>                 separators=separators,
>                 sort_keys=True,
>             )
> ```
> 
> hmm... I wonder if we really want this.

That's interesting. The order of the dictionary roles was not changed in my case, but it could be due to the role names.

**Comment by joshuagl** [comment]
> > I'm not totally sure why the targets dict order you chose isn't reflected in the serialized format
> 
> It's because we ask json to sort keys:
> 
> ```
>             json_bytes = json.dumps(
>                 metadata_obj.to_dict(),
>                 indent=indent,
>                 separators=separators,
>                 sort_keys=True,
>             )
> ```
> 
> hmm... I wonder if we really want this.

I think we probably do not. It's a bit of a trap if the serialisation format implementation _has_ to take care of sorting what is effectively input data to the serialisation methods. Conversely, if the serialisation format implementation does **not** have to take care of this, doing this during JSON serialisation/deserialisation could mask issues.

**Comment by MVrachev** [comment]
> > > I'm not totally sure why the targets dict order you chose isn't reflected in the serialized format
> > 
> > 
> > It's because we ask json to sort keys:
> > ```
> >             json_bytes = json.dumps(
> >                 metadata_obj.to_dict(),
> >                 indent=indent,
> >                 separators=separators,
> >                 sort_keys=True,
> >             )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > hmm... I wonder if we really want this.
> 
> I think we probably do not. It's a bit of a trap if the serialisation format implementation _has_ to take care of sorting what is effectively input data to the serialisation methods. Conversely, if the serialisation format implementation does **not** have to take care of this, doing this during JSON serialisation/deserialisation could mask issues.

I don't like the side effect of sorting the keys when calling `json_dumps()` creates. 
For example, if you read a target file and its delegated roles are not sorted by names _(keep in mind we store delegated roles as a dictionary where the key is the role name and the value is the role info)_, then when calling `Metadata.to_file()` the delegated roles will be sorted, but the signature won't be updated. I tested locally that the dictionaries order does make a difference when calculating this signature meaning the metadata signatures can no longer be used for verification.

You have to know to call `Metadata_sign()` beforehand which will sort the delegated roles as well (see https://github.com/secure-systems-lab/securesystemslib/blob/602d3fcd59f726fc64e78a93430f427712d26ca4/securesystemslib/formats.py#L646.) and then sign that data.

**Comment by jku** [comment]
> if you read a target file and its delegated roles are not sorted by names ..., then when calling Metadata.to_file() the delegated roles will be sorted

No. delegated roles is a list in json. It does not get sorted by json module.

**Comment by jku** [comment]
> It's a bit of a trap if the serialisation format implementation _has_ to take care of sorting what is effectively input data to the serialisation methods. Conversely, if the serialisation format implementation does **not** have to take care of this, doing this during JSON serialisation/deserialisation could mask issues.

I think this is in no way required: we just happen to do it that way. The reason for the sorting is likely this: it guarantees that Metadata objects that are "semantically" equal will produce the byte-for-byte same serialization. If the sorting wasn't done, then seemingly equal Metadata   objects can produce different json. Think of e.g. two equal Metadata objects where the same keys are added into root in a different order: this changes nothing from specification POV, but unless the keys are kept alpha-sorted, the output would be different...

The downside of this is that deserializing and immediately serializing metadata can lead to sort order changes (if the original metadata came from another implementation). It's probably not a major issue as we already do not guarantee byte-for-byte equivalence in that situation (even though we do guarantee that valid signatures stay valid).

**Comment by lukpueh** [comment]
I'm not sure if more comments are needed in this discussion, but let me see if we are all on the same page here:

1. The json standard is too vague to guarantee byte-for-byte deterministic serialized output for arbitrary json objects (see string encoding, whitespace, escape characters, decimal points, key order, etc.)
2. That's why we use a canonical json standard for signature computation/verification.
   1. Whether the specific canonical json spec we use is a good one, is a different questions (see 
       [sslib#159](https://github.com/secure-systems-lab/securesystemslib/issues/159))
   2. Whether parsing (canonical) json before verifying signatures is a good idea in general, is also a different question (e.g. the [dsse](https://github.com/secure-systems-lab/dsse) spec, which is currently developed as an alternative in our bubble, says no)
3. Because of (1) we can never guarantee that, if we deserialize an arbitrary json byte stream and re-serialize it, the output will be equal to the input. This is not a problem for signatures because auf (2), and it's also not a problem for verifying hashes of full metadata files  (i.e. targets in snapshot, snapshot in timestamp) because the exact serialized bytes used to compute the hash should always be available, when we need to verify that hash.
4. Regarding `sort_keys=True` in our json serializer, I think, due to known and limited set of values they can have, it actually makes the json objects we generate and serialize in python-tuf byte-for-byte deterministic, at least for a given Python/json version. We shouldn't rely on it, or make any guarantees to users, but I don't think it's an issue either.

**Comment by jku** [comment]
Yes, I agree with that list.

So I think this issue can be closed? Martin, let us know if you think there's something not covered above (or if the documentation needs to be improved)

**Comment by MVrachev** [comment]
I think I understand what's happening now and I can summarize why it was strange for me in the following points:
1. I hadn't paid attention before to the `securesystemslib.formats.encode_canonical()` and the need for the `CanonicalJSONSerializer` before. Now I see how we deterministically modify the metadata data before signing it, so that no matter how many whitespaces were used you always get the same signature. 
2.  After a discussion with Jussi it became clear why we are sorting our metadata when calling `JsonSerializer.serialize()`. It's understood that it's very likely that by deserializing and serializing a user won't get byte by byte equality (because of 1st point of Lukas comment),  then at least we can make sure that when we serialize to a dictionary the JSON keys will be sorted in a consistent way. That's how when we look into our test files we would know that `signatures` are first, `signed` is second, etc.
3. I verified that the insertion order is preserved when deserializing and serializing to a file. That's especially important for `Targets.delegations.roles` where the order makes a difference.
4. I answered my question on whether `different orders in the dictionaries in the metadata objects lead to a different signature?` The **answer** is yes, it does lead to a different signature.  Signing takes into account the content and the order of the data which is what we want especially for `Targets.delegations.roles` were changing a role position should lead to a new version.

I think I answered my questions.
If any of my points are wrong please correct me.
Thanks, @jku, @lukpueh and @joshuagl for reading this long script and helping me figure this out.

**Comment by MVrachev** [comment]
I am closing it for now.
If a new discussion forms, I will reopen it.
